From 673a6063d769ae0aed28efa2d17f054b29ffe93d Mon Sep 17 00:00:00 2001
From: Kris Shannon <k.shannon@amaze.com.au>
Date: Thu, 25 May 2017 10:07:14 +1000
Subject: [PATCH] Escape directory for moray filter in readdir

If the directory name contains any of the special LDAP filter characters - '(', ')', '*', or '=' - they need to be escaped before being added to the moray filter string.
---
 lib/common.js | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/lib/common.js b/lib/common.js
index b2d72ec..d2ee656 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -607,15 +607,16 @@ function readdir(dir, req) {
                   req.params.directory !== undefined);
     var hasObj = (req.params.obj !== undefined ||
                   req.params.object !== undefined);
+    var escDir = dir.replace(/[\\\(\)\=\*]/g, '\\$&');
 
     if ((hasDir && hasObj) || !(hasDir || hasObj)) {
-        filter += sprintf('(owner=%s)(dirname=%s)', account, dir);
+        filter += sprintf('(owner=%s)(dirname=%s)', account, escDir);
     } else if (hasDir) {
         filter += sprintf('(owner=%s)(dirname=%s)(type=directory)',
-                          account, dir);
+                          account, escDir);
     } else {
         filter += sprintf('(owner=%s)(dirname=%s)(type=object)',
-                          account, dir);
+                          account, escDir);
     }
 
     var marker = req.params.marker;
-- 
2.21.0

